compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -82,26 +93,261 @@ impl<'tcx> MirPass<'tcx> for InstrumentCoverage {
+        let is_fn_like = FnLikeNode::from_node(tcx.hir().get(hir_id)).is_some();
+
+        // Only instrument functions, methods, and closures (not constants since they are evaluated
+        // at compile time by Miri).
+        if !is_fn_like {
+            trace!("InstrumentCoverage skipped for {:?} (not an FnLikeNode)", mir_source.def_id());
+            return;
+        }
+        // FIXME(richkadel): By comparison, the MIR pass `ConstProp` includes associated constants,
+        // with functions, methods, and closures. I assume Miri is used for associated constants as
+        // well. If not, we may need to include them here too.
+
+        trace!("InstrumentCoverage starting for {:?}", mir_source.def_id());
+        Instrumentor::new(&self.name(), tcx, mir_source, mir_body).inject_counters();
+        trace!("InstrumentCoverage starting for {:?}", mir_source.def_id());
     }
 }
 
-#[derive(Clone)]
-struct CoverageRegion {
-    pub span: Span,
+/// A BasicCoverageBlock (BCB) represents the maximal-length sequence of CFG (MIR) BasicBlocks
+/// without conditional branches.
+#[derive(Debug, Clone)]
+struct BasicCoverageBlock {
     pub blocks: Vec<BasicBlock>,
 }
 
+impl BasicCoverageBlock {
+    pub fn leader_bb(&self) -> BasicBlock {
+        self.blocks[0]
+    }
+
+    pub fn id(&self) -> String {
+        format!(
+            "@{}",
+            self.blocks
+                .iter()
+                .map(|bb| bb.index().to_string())
+                .collect::<Vec<_>>()
+                .join(ID_SEPARATOR)
+        )
+    }
+}
+
+struct BasicCoverageBlocks {
+    vec: IndexVec<BasicBlock, Option<BasicCoverageBlock>>,
+}
+
+impl BasicCoverageBlocks {
+    pub fn from_mir(mir_body: &mir::Body<'tcx>) -> Self {
+        let mut basic_coverage_blocks =
+            BasicCoverageBlocks { vec: IndexVec::from_elem_n(None, mir_body.basic_blocks().len()) };
+        basic_coverage_blocks.extract_from_mir(mir_body);
+        basic_coverage_blocks
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &BasicCoverageBlock> {
+        self.vec.iter().filter_map(|option| option.as_ref())
+    }
+
+    fn extract_from_mir(&mut self, mir_body: &mir::Body<'tcx>) {
+        // Traverse the CFG but ignore anything following an `unwind`
+        let cfg_without_unwind = ShortCircuitPreorder::new(mir_body, |term_kind| {
+            let mut successors = term_kind.successors();
+            match &term_kind {
+                // SwitchInt successors are never unwind, and all of them should be traversed
+                TerminatorKind::SwitchInt { .. } => successors,
+                // For all other kinds, return only the first successor, if any, and ignore unwinds
+                _ => successors.next().into_iter().chain(&[]),
+            }
+        });
+
+        let mut blocks = Vec::new();
+        for (bb, data) in cfg_without_unwind
+            .filter(|(_, data)| data.terminator().kind != TerminatorKind::Unreachable)
```

Comment:
If we were to insert counters in `Unreachable` blocks, they would never be reached, so there should be no difference in the output, but we might end up inserting LLVM instrprof counters in the code for no reason. Generally this is an easy performance optimization.

From what I've seen, `Unreachable` blocks exist outside (disconnected from) the main CFG.

Reply:
> From what I've seen, Unreachable blocks exist outside (disconnected from) the main CFG.

If that's the case then you shouldn't need to filter them explicitly. The MIR traversal already ignores blocks that aren't connected to the CFG.

Reply:
I believe you're right. Maybe I added the filter because I wasn't sure at the time, rather than having observed otherwise. I've seen MIR's with Unreachable terminators in recent test programs, but I'm having a hard time reproducing that situation, so I can't confirm this for sure. But looking at the Preorder code, it must be true.

I'll remove the filter and push the change as soon as I re-run the tests.

Reply:
Done.